discovery/watcher.go

```diff
@@ -0,0 +1,339 @@
+
+type Watcher struct {
+	config Config
+	log    hclog.Logger
+
+	currentServer atomic.Value
+
+	initComplete atomic.Value
+	conn         *grpc.ClientConn
+	resolver     *watcherResolver
+	token        string
+
+	backoff backoff.BackOff
+
+	once sync.Once
+
+	// interface to allow us to inject custom server ports for tests
+	discoverer Discoverer
+}
+
+type serverState struct {
+	addr              Addr
+	dataplaneFeatures map[string]bool
+}
+
+func NewWatcher(config Config, log hclog.Logger) *Watcher {
+	if log == nil {
+		log = hclog.NewNullLogger()
+	}
+
+	// TODO: config for backoff values
+	backoff := backoff.NewExponentialBackOff()
+	backoff.MaxElapsedTime = 0 // Allow backing off forever.
+
+	w := &Watcher{
+		config:     config,
+		log:        log,
+		backoff:    backoff,
+		resolver:   newResolver(log),
+		discoverer: NewNetaddrsDiscoverer(config, log),
+	}
+	w.initComplete.Store(false)
+	w.currentServer.Store(serverState{})
+	return w
+}
+
+func (w *Watcher) Subscribe() chan SubscribeState {
+	// TODO: add this
+	panic("unimplemented")
+}
+
+// Run spawns a goroutine to watch for Consul server set changes.
+// It blocks to wait for initialization to complete, and returns
+// initial state or an error on failure.
+func (w *Watcher) Run(ctx context.Context) (*InitState, error) {
+	if w.conn == nil {
+		var cred credentials.TransportCredentials
+		if tls := w.config.TLS; tls != nil {
+			cred = credentials.NewTLS(tls)
+		} else {
+			cred = insecure.NewCredentials()
+		}
+
+		dialOpts := []grpc.DialOption{
+			grpc.WithTransportCredentials(cred),
+			grpc.WithResolvers(w.resolver), // note: experimental api.
+			// TODO: Add interceptors here.
+			// TODO: Add custom grpc balancer here.
+		}
+
+		// Dial with "consul://" to trigger our custom resolver. We don't
+		// provide a server address. The connection will be updated by the
+		// Watcher via the custom resolver once an address is known.
+		conn, err := grpc.DialContext(ctx, "consul://", dialOpts...)
+		if err != nil {
+			return nil, err
+		}
+		w.conn = conn
+	}
+
+	// Spawn our goroutine.
+	w.once.Do(func() { go w.run(ctx) })
+
+	// Wait for init to complete.
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case <-time.After(200 * time.Millisecond):
+		}
+
+		if w.initComplete.Load().(bool) {
+			break
+		}
+	}
+
+	current := w.currentServer.Load().(serverState)
+	return &InitState{
+		GRPCConn:          w.conn,
+		Token:             w.token,
+		Address:           current.addr,
+		DataplaneFeatures: current.dataplaneFeatures,
+	}, nil
+}
+
+func (w *Watcher) run(ctx context.Context) {
+	// addrs is the current set of servers we know about.
+	var addrs *addrSet
+	var err error
+
+	for {
+		// Find and connect to a server.
+		//
+		// When this successfully connects to a server, it returns the chosen
+		// server and the latest set of server addresses. If we get an error,
+		// then we retry with backoff.
+		addrs, err = w.nextServer(ctx, addrs)
```

Comment:
I should clarify that `nextServer` will block after "connecting" to a server. It will open the server watch stream, and wait for changes from the stream. If the server watch stream is disabled, then it will sleep until there is some indicator it should change servers.